crates/sui-client/src/lib.rs

```diff
@@ -0,0 +1,363 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::apis::{
+    RpcBcsApi, RpcFullNodeReadApi, RpcGatewayApi, RpcReadApi, RpcTransactionBuilder, WalletSyncApi,
+};
+
+use async_trait::async_trait;
+use jsonrpsee::http_client::{HttpClient, HttpClientBuilder};
+use jsonrpsee::ws_client::{WsClient, WsClientBuilder};
+use sui_json::SuiJsonValue;
+use sui_json_rpc::api::RpcBcsApiClient;
+use sui_json_rpc::api::RpcFullNodeReadApiClient;
+use sui_json_rpc::api::RpcGatewayApiClient;
+use sui_json_rpc::api::RpcReadApiClient;
+use sui_json_rpc::api::RpcTransactionBuilderClient;
+use sui_json_rpc::api::WalletSyncApiClient;
+use sui_json_rpc_types::{
+    GatewayTxSeqNumber, GetObjectDataResponse, GetRawObjectDataResponse,
+    RPCTransactionRequestParams, SuiObjectInfo, SuiTypeTag, TransactionBytes,
+    TransactionEffectsResponse, TransactionResponse,
+};
+use sui_types::base_types::{ObjectID, SuiAddress, TransactionDigest};
+use sui_types::sui_serde::Base64;
+
+pub mod apis;
+pub mod crypto;
+
+// re-export essential sui crates
+pub use sui_json as json;
+pub use sui_types as types;
+
+pub struct SuiRpcClient {
+    client: Client,
+}
+
+impl SuiRpcClient {
+    pub fn new_http_client(server_url: &str) -> Result<Self, anyhow::Error> {
+        let client = HttpClientBuilder::default().build(server_url)?;
+        Ok(Self {
+            client: Client::Http(client),
+        })
+    }
+
+    pub async fn new_ws_client(server_url: &str) -> Result<Self, anyhow::Error> {
+        let client = WsClientBuilder::default().build(server_url).await?;
+        Ok(Self {
+            client: Client::Ws(client),
+        })
+    }
+}
+
+enum Client {
+    Http(HttpClient),
+    Ws(WsClient),
+}
+
+#[async_trait]
+impl RpcReadApi for SuiRpcClient {
+    async fn get_objects_owned_by_address(
+        &self,
+        address: SuiAddress,
+    ) -> anyhow::Result<Vec<SuiObjectInfo>> {
+        Ok(match &self.client {
+            Client::Http(c) => c.get_objects_owned_by_address(address),
+            Client::Ws(c) => c.get_objects_owned_by_address(address),
+        }
+        .await?)
+    }
```

Comment:
Actually here's my recommendation: Completely remove the extra copies of these traits and instead of implementing these traits on `SuiRpcClient` just directly add the methods you have implemented here in a single impl block:

```
impl SuiRpcClient {
    async fn get_objects_owned_by_address(..)
...
}
```

Reply:
This will actually make the client more usable since now you wont need to bring into scope these traits to actually use these methods.

Reply:
Let me know what you think, i may be missing something else